packages/p2p/src/components/my-ads/quick-add-modal.jsx

```diff
@@ -75,491 +42,363 @@ const QuickAddModal = ({ advert }) => {
         }
     };
 
-    const setShouldCloseAllModals = should_close_all_modals => {
-        setSelectedMethods([]);
-        if (!should_close_all_modals) {
-            if (my_ads_store.should_show_add_payment_method) {
+    const is_payment_methods_selected =
+        my_profile_store.is_add_payment_method_form_modified || (is_buy_advert && selected_methods.length > 0);
+
+    const closeModals = should_close_all => {
+        if (is_payment_methods_selected) {
+            my_profile_store.showCancelAddPaymentMethodModal();
+        } else {
+            clearSelectedPaymentMethods();
+            if (is_buy_advert || should_close_all) {
                 my_ads_store.setShouldShowAddPaymentMethod(false);
-            } else {
                 my_ads_store.hideQuickAddModal();
+            } else {
+                if (!my_ads_store.should_show_add_payment_method) {
+                    my_ads_store.hideQuickAddModal();
+                }
+                my_ads_store.setShouldShowAddPaymentMethod(false);
             }
-        } else {
-            my_ads_store.setShouldShowAddPaymentMethod(false);
-            my_ads_store.hideQuickAddModal();
         }
     };
 
-    React.useEffect(() => {
+    const toggleModal = e => {
+        if (
+            !e.target ||
+            ['svg', 'use'].includes(e.target.tagName) ||
+            ['dc-dropdown-list__item', 'dc-btn'].every(className => !e.target.className.includes(className))
+        ) {
+            closeModals(true);
+        }
+    };
+
+    const clearSelectedPaymentMethods = () => {
```

Comment:
could we move these functions into the store? the quick add modal is getting quite cluttered